core/src/main/java/org/apache/iceberg/util/ExpireSnapshotUtil.java

```diff
@@ -0,0 +1,397 @@
+    ManifestExpirationChanges manifestExpirationChanges =
+        findExpiredManifestsInUnusedSnapshots(snapshotsNotChangingTableState, validManifests,
+            currentTableMetadata, expiredIds, ops);
+
+    manifestExpirationChanges.getManifestsToScan().addAll(manifestsToScan);
+    return manifestExpirationChanges;
+  }
+
+  public static SnapshotExpirationChanges getExpiredSnapshots(
+      TableOperations ops, TableMetadata originalMetadata) {
+
+    TableMetadata current = ops.refresh();
+
+    Set<Long> validIds = Sets.newHashSet();
+    for (Snapshot snapshot : current.snapshots()) {
+      validIds.add(snapshot.snapshotId());
+    }
+
+    Set<Long> expiredIds = Sets.newHashSet();
+    for (Snapshot snapshot : originalMetadata.snapshots()) {
+      long snapshotId = snapshot.snapshotId();
+      if (!validIds.contains(snapshotId)) {
+        // the snapshot was expired
+        LOG.info("Expired snapshot: {}", snapshot);
+        expiredIds.add(snapshotId);
+      }
+    }
+
+    return new SnapshotExpirationChanges(current.snapshots(), validIds, expiredIds);
+  }
+
+  public static class SnapshotExpirationChanges {
+
+    private final List<Snapshot> currentSnapshots;
+    private final Set<Long> validSnapshotIds;
+    private final Set<Long> expiredSnapshotIds;
+
+    public SnapshotExpirationChanges(
+        List<Snapshot> currentSnapshots, Set<Long> validSnapshotIds, Set<Long> expiredSnapshotIds) {
+
+      this.currentSnapshots = currentSnapshots;
+      this.validSnapshotIds = validSnapshotIds;
+      this.expiredSnapshotIds = expiredSnapshotIds;
+    }
+
+    public List<Snapshot> getCurrentSnapshots() {
+      return currentSnapshots;
+    }
+
+    public Set<Long> getValidSnapshotIds() {
+      return validSnapshotIds;
+    }
+
+    public Set<Long> getExpiredSnapshotIds() {
+      return expiredSnapshotIds;
+    }
+  }
+
+  public static class ManifestExpirationChanges {
+
+    private final Set<ManifestFile> manifestsToScan;
+    private final Set<ManifestFile> manifestsToRevert;
+    private final Set<String> manifestsToDelete;
+    private final Set<String> manifestListsToDelete;
+
+    private ManifestExpirationChanges(
+        Set<ManifestFile> manifestsToScan, Set<ManifestFile> manifestsToRevert,
+        Set<String> manifestsToDelete, Set<String> manifestListsToDelete) {
+
+      this.manifestsToScan = manifestsToScan;
+      this.manifestsToRevert = manifestsToRevert;
+      this.manifestsToDelete = manifestsToDelete;
+      this.manifestListsToDelete = manifestListsToDelete;
+    }
+
+
+    public Set<ManifestFile> getManifestsToScan() {
+      return manifestsToScan;
+    }
+
+    public Set<ManifestFile> getManifestsToRevert() {
+      return manifestsToRevert;
+    }
+
+    public Set<String> getManifestsToDelete() {
+      return manifestsToDelete;
+    }
+
+    public Set<String> getManifestListsToDelete() {
+      return manifestListsToDelete;
+    }
+  }
+
+  private static class AncestorIds {
```

Comment:
We may need to play around with the name as it is not simply ancestor ids, it is used for cherry-picked changes.

Reply:
ActiveSnapshots? AncestorInfo ?

Reply:
Actually, we may just have a method that computes `pickedAncestorSnapshotIds` that could accept `ancestorIds` as a param. Computing `ancestorIds` is one line. Then we won't need the wrapper and can return `Set<Long>`.